awx/main/scheduler/task_manager_models.py

```diff
@@ -33,9 +33,55 @@ def remaining_capacity(self):
         return remaining
 
 
+class TaskManagerInstanceGroup:
+    """A class representing minimal data the task manager needs to represent an InstanceGroup."""
+
+    def __init__(self, obj, task_manager_instances=None):
+        if not isinstance(task_manager_instances, TaskManagerInstances):
+            raise TypeError("TaskManagerInstanceGroup must be initialized with an instance of TaskManagerInstances as an argument")
+        self.name = obj.name
+        self.is_container_group = obj.is_container_group
+        self.running_jobs = task_manager_instances.instance_groups_running_jobs.get(self.name, 0)
+        self.consumed_forks = task_manager_instances.instance_groups_consumed_forks.get(self.name, 0)
+        _instances = obj.instances.all()
+        # This is somewhat paranoid logic we've inherited from the past that is concerned with ensuring we filter out any instances that
+        # may have just been added to the instance group in the time between when we built the TaskManagerInstances and are building this object
+        self.instances = [task_manager_instances[instance.hostname] for instance in _instances if instance.hostname in task_manager_instances]
+        self.instance_hostnames = tuple([instance.hostname for instance in _instances if instance.hostname in task_manager_instances])
+        self.max_concurrent_jobs = obj.max_concurrent_jobs
+        self.max_forks = obj.max_forks
+
+    def consume_capacity(self, task=None, control_impact=False):
+        if control_impact:
+            self.consumed_forks += settings.AWX_CONTROL_NODE_TASK_IMPACT
+            return
+        self.running_jobs += 1
+        if task and (task.execution_node in self.instance_hostnames or self.is_container_group):
+            self.consumed_forks += task.task_impact
```

Comment:
Ok, just remembered why this never worked in the old task manager, which tried to track this (but for no reason)

Instances can be members of multiple instance groups

This implementation is naive, we consume capacity when the instance group is assigned.
For container groups, this will always work, since they don't have instances.
But for regular instance groups, what about when an instance is in multiple instance groups? Do we have a reverse mapping of `instance: list of instance groups` and consume capacity on all of them? Suppose I could do that

Reply:
Resolved this now